### PR TITLE
Add autograd for to_sparse.

### PR DESCRIPTION
**Pull request by nairbv**
https://github.com/pytorch/pytorch/issues/18111

### CODE REVIEW

**Comment by nairbv** [comment]
As discussed with @eellison, need feedback from jit team on support for sparse tensors.

**Comment by eellison** [comment]
I don't think we want to add the additional complexity of Sparse tensors to JIT yet. 

We have an existing pass `inpace_check.cpp` that checks for one particular unsupported op before running the JIT. Maybe you could rename the pass and check for `aten::to_sparse` in that pass as well? 

How does that sound @suo @zdevito

File: torch/csrc/jit/ir.h [hunk 1]
```diff
@@ -163,7 +163,8 @@ struct Value {
  public:
   Value* setType(TypePtr type);
   void inferTypeFrom(const at::Tensor& output) {
-    setType(CompleteTensorType::create(output));
+    auto complete = CompleteTensorType::create(output.scalar_type(), output.options().device(), output.sizes());
```
Comment (zdevito) [inline torch/csrc/jit/ir.h]: Why did this change? The details of how to create a CompleteTensorType from Tensor need to stay local to where it is defined, otherwise it is hard to update.

File: torch/csrc/jit/pybind_utils.h [hunk 2]
```diff
@@ -91,10 +91,8 @@ inline c10::optional<TypePtr> unifyOrInitializeType(
 inline TypedIValue toTypedIValue(py::handle input) {
   if (THPVariable_Check(input.ptr())) {
     auto ten = py::cast<at::Tensor>(input);
-    if (ten.is_sparse()) {
-      AT_ERROR("sparse tensors not supported");
-    }
-    return TypedIValue(ten, CompleteTensorType::create(ten));
+    auto complete = CompleteTensorType::create(ten.scalar_type(), ten.device(), ten.sizes());
```
Comment (zdevito) [inline torch/csrc/jit/pybind_utils.h]: Also leaking the details of how to create a CompleteTensorType. This will be nearly impossible to find if we change complete tensor type.

File: torch/csrc/jit/pybind_utils.h [hunk 3]
```diff
@@ -220,9 +218,6 @@ inline IValue toIValue(
     case TypeKind::ProfiledTensorType:
     case TypeKind::CompleteTensorType: {
       auto var = py::cast<autograd::Variable>(obj);
-      if (var.is_sparse()) {
```
Comment (zdevito) [inline torch/csrc/jit/pybind_utils.h]: We can't delete these without more work to figure out what is needed to support sparse tensors in the JIT.

File: torch/csrc/jit/ir.h (hunk 1 above)
Comment (nairbv) [reply torch/csrc/jit/ir.h]: Just because if the default create() function calls strides() on a sparse tensor it'll fail.

File: torch/csrc/jit/pybind_utils.h (hunk 3 above)
Comment (nairbv) [reply torch/csrc/jit/pybind_utils.h]: reverted jit changes and excluded to_sparse from jit tests instead.

Is there a plan to support sparse tensors in jit at some point?

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -938,8 +941,8 @@
   self: zeros_like(grad)
 
 - name: sparse_mask(Tensor self, Tensor mask)
-  self: not_implemented("sparse_mask")
-  mask: not_implemented("sparse_mask")
+  self: grad.to_dense().sparse_mask(mask).to_dense()
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml]: Why is this necessary to implement `to_sparse` backward?
Also, this (`sparse_mask` backward) needs a test given that you've implemented it.

File: test/test_jit.py [hunk 5]
```diff
@@ -14961,7 +14965,8 @@ def check(name):
                 args_tensor = deepcopy(unpack_variables(args_variable))
 
                 def fn(*inputs, **kwargs):
-                    output = getattr(inputs[0], name)(*inputs[1:], **kwargs)
+                    attr = getattr(inputs[0], name)
```
Comment (zou3519) [inline test/test_jit.py]: nit: this isn't necessary, but it's fine to keep it

File: test/test_jit.py [hunk 6]
```diff
@@ -14153,6 +14155,8 @@ def forward(self, x, y):
 
     # unknown builtin op
     'test_nn_fold',
+
+    'test_to_sparse'
```
Comment (zou3519) [inline test/test_jit.py]: You should probably list why this is here "# jit doesn't support sparse tensors."

File: test/test_autograd.py [hunk 7]
```diff
@@ -1624,6 +1624,13 @@ def test_sparse_gather_x_scalar(self):
     def test_sparse_gather_both_scalar(self):
         self._test_sparse_gather((), (), 0)
 
+    def test_to_sparse_autograd(self):
```
Comment (zou3519) [inline test/test_autograd.py]: Why is this function necessary? There's already a to_sparse test in common_method_invocations that autograd should call

**Review by zou3519 — COMMENTED** [review]
logic looks fine, some comments though

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (nairbv) [reply tools/autograd/derivatives.yaml]: in `./aten/src/ATen/native/TensorConversions.cpp` you can see that to_dense_backward calls sparse_mask, which ends up causing this to be required.

File: test/test_jit.py (hunk 5 above)
Comment (nairbv) [reply test/test_jit.py]: I just broke this up to make it easier to add print statements when I was trying to figure out how to get sparse to work in jit.

File: test/test_autograd.py (hunk 7 above)
Comment (nairbv) [reply test/test_autograd.py]: probably not, I added this first before realizing I should add in common method invocs.

File: torch/autograd/gradcheck.py [hunk 8]
```diff
@@ -307,7 +307,7 @@ def fn(input):
                 continue
             if isinstance(gi, torch.Tensor) and gi.layout != torch.strided:
                 if gi.layout != i.layout:
-                    return fail_test('grad is incorrect layout')
+                    return fail_test('grad is incorrect layout (' + str(gi.layout) + ' ne ' + str(i.layout) + ')')
```
[the remark this answers is not in the document]
Comment (nairbv) [reply torch/autograd/gradcheck.py]: sure will do

File: test/test_autograd.py (hunk 7 above)
Comment (nairbv) [reply test/test_autograd.py]: I wasn't able to add a sparse mask test in common invocs because it's not setup to handle tests with sparse inputs, so I converted this test to test sparse_mask.

does sparse mask also gets tested indirectly via the common method invocations entry for to_sparse?

File: test/test_autograd.py [hunk 9]
```diff
@@ -1624,6 +1624,17 @@ def test_sparse_gather_x_scalar(self):
     def test_sparse_gather_both_scalar(self):
         self._test_sparse_gather((), (), 0)
 
+    # autograd tests via common_method_invocations don't allow input tensors to
+    # be sparse (RuntimeError: gradcheck expects all tensor inputs are dense when
+    # check_sparse_nnz is set to False.)
+    def test_sparse_mask_autograd(self):
+        for device in ['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']:
+            tensor = torch.randn(3, requires_grad=True, device=device)
+            mask = torch.ones(3, device=device).to_sparse()
```
Comment (zou3519) [inline test/test_autograd.py]: You should test a mask that is not just all ones, something like torch.tensor([1, 0, 1])

File: test/test_autograd.py (hunk 7 above)
Comment (zou3519) [reply test/test_autograd.py]: gradcheck and gradgradcheck are run via the common method invocations, so yes sparse_mask is indirectly tested (in the gradgradcheck step), but it's always good to have an explicit test

File: test/test_autograd.py (hunk 9 above)
Comment (nairbv) [reply test/test_autograd.py]: ah yes, that was a bit lazy of me. fixed.

**Review by zou3519 — APPROVED** [review]
lgtm

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (gchanan) [reply tools/autograd/derivatives.yaml]: why do you need to densify the grad?  Modulo dealing with coalescing (does this do that correctly?), isn't the just creating a sparse tensor with the values of grad correct?

File: test/test_autograd.py [hunk 10]
```diff
@@ -1624,6 +1624,19 @@ def test_sparse_gather_x_scalar(self):
...
+    # check_sparse_nnz is set to False.)
+    def test_sparse_mask_autograd(self):
+        for device in ['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']:
+            tensor = torch.randn(3, requires_grad=True, device=device)
+            mask = torch.ones(3, device=device)
+            mask[1] = 0
+            mask = mask.to_sparse()
+            converted = tensor.sparse_mask(mask).to_dense()
+            converted.sum().backward()
+            self.assertEqual(tensor.grad, mask.to_dense())
```
Comment (gchanan) [inline test/test_autograd.py]: does gradcheck / gradgradcheck not work here?

Comment (zou3519) [reply test/test_autograd.py]: I think @nairbv said gradcheck / gradgradcheck can't handle sparse tensors (as autograd leafs).

But it could be good to run gradcheck on `tensor.sparse_mask(mask.to_sparse())` (where `tensor`, `mask` are dense tensors, if that works.

Comment (gchanan) [reply test/test_autograd.py]: ya, I think we've run into this before and it can be worked around.

Comment (nairbv) [reply test/test_autograd.py]: sparse_mask didn't allow for a dense mask

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (nairbv) [reply tools/autograd/derivatives.yaml]: sparse_mask requires a dense tensor and sparse mask. if I don't call to_dense() on the grad, iirc, sparse mask rejects it. If I don't call to_dense on the result, IIRC I got an error that the gradient layout has to match. I guess we could write an explicit kernel that took a sparse gradient and returned a dense mask though. I can add one, or at least file a ticket if you think it's worthwhile?
